File: apps/cli/src/commands/report.ts
/**
 * Command-pattern implementation of the `report` CLI command.
 *
 * hem report [ref] [--out <path>] [--json] --project .
 *
 * Outputs a markdown report (to stdout or --out file).
 * Shows diff when comparing against a stored snapshot.
 */

import { writeFile } from "node:fs/promises";
import path from "node:path";

import { captureCurrentState } from "@qxinm/hem-core/current-state.js";
import { diffGraphs, type GraphDiff } from "@qxinm/hem-core/diff.js";
import { renderMarkdownReport } from "@qxinm/hem-core/report.js";
import { scanProject, type ScanResult } from "@qxinm/hem-core/scan.js";
import { readSnapshot } from "@qxinm/hem-core/store.js";
import type { Snapshot } from "@qxinm/hem-core/types.js";
import { hasFlag, json, runtimeOptions, valueAfter, type RuntimeOptions } from "../cli-shared.js";
import type { Command, CommandContext } from "./index.js";

async function snapshotByRef(ref: string, args: string[]): Promise<Snapshot> {
  if (ref === "current") {
    return (await captureCurrentState(runtimeOptions(args))).snapshot;
  }
  const opts = runtimeOptions(args);
  return await readSnapshot(opts.storeDir, ref, opts.agent);
}

function trustForReport(scan: ScanResult): { readOnly: boolean; network: "disabled"; commandsExecuted: number } {
  return {
    readOnly: scan.trust.readOnly,
    network: scan.trust.network,
    commandsExecuted: scan.trust.commandsExecuted.length
  };
}

// ── Command definition ───────────────────────────────────────────────────────

export const reportCommand: Command = {
  name: "report",
  description:
    "Generate a markdown report of agent state, findings, and (optionally) diff. " +
    "Usage: hem report [ref] [--out <path>] [--json] --project .",

  async execute(ctx: CommandContext): Promise<number> {
    const { args } = ctx;

    // ref is the first positional argument; default to "current"
    const ref = args[1] ?? "current";
    const options: RuntimeOptions = runtimeOptions(args);
    const snapshot = await snapshotByRef(ref, args);

    // Compute diff only when comparing against a stored snapshot
    const diff: GraphDiff | undefined =
      ref === "current"
        ? undefined
        : diffGraphs(snapshot.graph, (await captureCurrentState(options)).snapshot.graph);

    // Scan is needed for trust info and blind spots
    const scan: ScanResult =
      ref === "current"
        ? (await captureCurrentState(options)).scan
        : await scanProject(options);

    const markdown = renderMarkdownReport({
      snapshotName: snapshot.manifest.name,
      trust: trustForReport(scan),
      evidence: snapshot.evidence,
      graph: snapshot.graph,
      findings: snapshot.auditFindings,
      provenance: snapshot.provenance,
      blindSpots: scan.blindSpots,
      diffs: diff
    });

    if (hasFlag(args, "--json")) {
      process.stdout.write(json({ snapshot, markdown }));
      return 0;
    }

    const out = valueAfter(args, "--out");
    if (out) {
      await writeFile(path.resolve(out), markdown);
      process.stdout.write(`Wrote report: ${path.resolve(out)}\n`);
    } else {
      process.stdout.write(markdown);
    }

    return 0;
  }
};

File: internal/cli/report.go
package cli

import (
	"os"

	"github.com/qyinm/hem/internal/hemcore/diff"
	"github.com/qyinm/hem/internal/hemcore/report"
	"github.com/qyinm/hem/internal/hemcore/scan"
	"github.com/qyinm/hem/internal/hemcore/snapshot"
	"github.com/qyinm/hem/internal/hemcore/types"
	"github.com/spf13/cobra"
)

func newReportCmd() *cobra.Command {
	var common CommonFlags
	var out string

	cmd := &cobra.Command{
		Use:   "report [reference]",
		Short: "Generate a markdown report of agent state and findings",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			reference := "current"
			if len(args) > 0 {
				reference = args[0]
			}
			exitCode := runReport(cmd, &common, reference, out)
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	cmd.Flags().StringVar(&out, "out", "", "Write report to file instead of stdout")
	return cmd
}

func runReport(cmd *cobra.Command, common *CommonFlags, reference, out string) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	snap, snapErr := snapshotByRef(reference, &runtime)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	var graphDiff *diff.GraphDiff
	if reference != "current" {
		current, err := snapshot.CaptureCurrentState(&runtime, "current")
		if err != nil {
			return writeError(cmd.ErrOrStderr(), &types.SnapError{
				Code:    "HEM_CURRENT_STATE_FAILED",
				Problem: "Failed to capture current state.",
				Cause:   err.Error(),
				Fix:     "Verify project and store paths are accessible.",
			})
		}
		d := diff.DiffGraphs(snap.Graph, current.Snapshot.Graph)
		graphDiff = &d
	}

	var scanResult types.ScanResult
	if reference == "current" {
		current, err := snapshot.CaptureCurrentState(&runtime, "current")
		if err != nil {
			return writeError(cmd.ErrOrStderr(), &types.SnapError{
				Code:    "HEM_CURRENT_STATE_FAILED",
				Problem: "Failed to capture current state.",
				Cause:   err.Error(),
				Fix:     "Verify project and store paths are accessible.",
			})
		}
		scanResult = current.Scan
	} else {
		scanResult = scan.ScanProject(&types.ScanOptions{
			ProjectPath: runtime.ProjectPath,
			HomeDir:     runtime.HomeDir,
			StoreDir:    runtime.StoreDir,
			Agent:       runtime.Agent,
			Scope:       runtime.Scope,
		})
	}

	snapshotName := snap.Manifest.Name
	markdown := report.RenderMarkdownReport(&report.Input{
		SnapshotName: &snapshotName,
		Trust: report.Trust{
			ReadOnly:         scanResult.Trust.ReadOnly,
			Network:          scanResult.Trust.Network,
			CommandsExecuted: uint32(len(scanResult.Trust.CommandsExecuted)),
		},
		Evidence:   snap.Evidence,
		Graph:      snap.Graph,
		Findings:   snap.AuditFindings,
		Provenance: snap.Provenance,
		BlindSpots: scanResult.BlindSpots,
		Diffs:      graphDiff,
	})

	if common.JSON {
		data, err := report.MarshalInputJSON(snap, markdown)
		if err != nil {
			return writeError(cmd.ErrOrStderr(), &types.SnapError{
				Code:    "HEM_JSON_SERIALIZE_FAILED",
				Problem: "Failed to serialize report JSON.",
				Cause:   err.Error(),
				Fix:     "This is an internal error.",
			})
		}
		if _, err := cmd.OutOrStdout().Write(append(data, '\n')); err != nil {
			return 1
		}
		return 0
	}

	if out != "" {
		if err := os.WriteFile(out, []byte(markdown), 0o644); err != nil {
			return writeError(cmd.ErrOrStderr(), &types.SnapError{
				Code:    "HEM_REPORT_WRITE_FAILED",
				Problem: "Failed to write report file.",
				Cause:   err.Error(),
				Fix:     "Verify the output path is writable.",
			})
		}
		return writeStdout(cmd.OutOrStdout(), "Report written to "+out+"\n")
	}
	return writeStdout(cmd.OutOrStdout(), markdown)
}